src/plugins/workspace/public/components/workspace_creator/utils.ts

```diff
@@ -0,0 +1,147 @@
+    for (const permission of permissions) {
+      if (permission.type === WorkspacePermissionItemType.User) {
+        userPermissions.push({ id: permission.userId, modes: permission.modes });
+      }
+      if (permission.type === WorkspacePermissionItemType.Group) {
+        groupPermissions.push({ id: permission.group, modes: permission.modes });
+      }
+    }
+  }
+  return [userPermissions, groupPermissions];
+};
+
+const getUnsavedUserOrGroupPermissionChangesCount = (
+  initialPermissions: UserOrGroupPermissionEditingData,
+  currentPermissions: UserOrGroupPermissionEditingData
+) => {
+  // for user or group permissions, unsaved changes is the sum of
+  // # deleted permissions, # added permissions and # edited permissions
+  let addedPermissions = 0;
+  let editedPermissions = 0;
+  const initialPermissionMap = new Map<string, WorkspacePermissionMode[]>();
+  for (const permission of initialPermissions) {
+    if (permission.id) {
+      initialPermissionMap.set(permission.id, permission.modes ?? []);
+    }
+  }
+
+  for (const permission of currentPermissions) {
+    if (!permission.id) {
+      addedPermissions += 1; // added permissions
+    } else {
+      const permissionModes = initialPermissionMap.get(permission.id);
+      if (!permissionModes) {
+        addedPermissions += 1;
+      } else if (
+        getPermissionModeId(permissionModes) !== getPermissionModeId(permission.modes ?? [])
+      ) {
+        editedPermissions += 1; // added or edited permissions
+      }
+    }
+  }
+
+  // currentPermissions.length = initialPermissions.length + # added permissions - # deleted permissions
+  const deletedPermissions =
+    addedPermissions + initialPermissions.length - currentPermissions.length;
+
+  return addedPermissions + editedPermissions + deletedPermissions;
+};
+
+export const getUnsavedChangesCount = (
+  initialFormData: WorkspaceFormEditingData,
+  currentFormData: WorkspaceFormEditingData
+) => {
+  let unsavedChangesCount = 0;
+  if (initialFormData.name !== currentFormData.name) {
+    unsavedChangesCount += 1;
+  }
+  // initial and current description could be undefined
+  const initialDescription = initialFormData.description ?? '';
+  const currentDescription = currentFormData.description ?? '';
+  if (initialDescription !== currentDescription) {
+    unsavedChangesCount += 1;
+  }
+  if (initialFormData.color !== currentFormData.color) {
+    unsavedChangesCount += 1;
+  }
+  if (initialFormData.icon !== currentFormData.icon) {
+    unsavedChangesCount += 1;
+  }
+  if (initialFormData.defaultVISTheme !== currentFormData.defaultVISTheme) {
+    unsavedChangesCount += 1;
+  }
+  const featureIntersectionCount = (
+    initialFormData.features?.filter((feature) => currentFormData.features?.includes(feature)) ?? []
+  ).length;
+  // for features, unsaved changes is the sum of # deleted features and # added features
+  unsavedChangesCount += (initialFormData.features?.length ?? 0) - featureIntersectionCount;
```

Comment:
Do we need to consider features with feature rule like `@category`?

Reply:
Thanks for pointing out. I will reimplement this function with **featureMatchesConfig**